### PR TITLE
docking tests: code in plots and monitors should run and be checked

### PR DESCRIPTION
**Pull request by mrerrormessage**

### CODE REVIEW

**Comment by mrerrormessage** [comment]
As I was working on this, I noticed that generating a model is done in a somewhat piecemeal fashion. Each of several classes adds their bit in, and they're all concatenated and written.
This system has worked effectively so far, but is getting a bit gnarly. I was wondering whether it might be better to have the model split into sections, into which each piece could add bits.
It seems to me that the model is split into a few different parts, and different pieces touch different parts:
* Required javascripts
* Workspace dependency creation (limited to plots at the moment, but easy to imagine more uses)
* Workspace initialization
* Top-Level variable creation (procedures, widgets, and workspace fields)
So after this change, we would have something like the following (which is only a very rough sketch)
```scala
trait ModelGenerator {
  def requires: String
  def preWorkspaceInit: String
  def workspaceInit: String
  def postWorkspaceInit: String
}

def writeSection(mgs: Seq[ModelGenerator], f: ModelGenerator => String): String =
  mgs.map(f).mkString("\n")

def writeModel: String = {
  val modelComponents = Seq(ModelConfigGenerator, RuntimeInit, WidgetCompilation)                                                                                                                writeSection(modelComponents, _.requires)                                                                                                                                                      writeSection(modelComponents, _.preWorkspaceInit)                                                                                                                                              writeSection(modelComponents, _.workspaceInit)                                                                                                                                                 writeSection(modelComponents, _.postWorkspaceInit)                                                                                                                                           }                                                                                                                                                                                              ```                                                                                                                                                               
I was tempted to add in this refactor as part of this PR, but I held off because this seemed like an change worthy of discussion and because the scope was really outside of this particular issue.
I did, however, want to note the idea because we're starting to get to the point of having a lot of model components, and this seemed like a way to simplify some of what we have.
If you would like to see a change similar to this one, I'd be happy to open an issue.

File: jvm/src/test/scala/JavascriptObjectTest.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    val javascriptObject = new JavascriptObject("baz" -> jsFunction).addObjectProperties(jsObject)
+    assertResult(Some(Right(jsFunction)))(javascriptObject("baz"))
+    assertResult(Some(Left(JsString("bar"))))(javascriptObject("foo"))
+  }
+
+  test("asJsonObject returns a JsObject instead of a string, with function values excluded") {
+    val javascriptObject = new JavascriptObject(jsObject)
+    assertResult(jsObject)(javascriptObject.asJsonObject)
+  }
+  test("asJsonObject return a JsObject with all jsObjects merged") {
```
Comment (TheBizzle) [inline jvm/src/test/scala/JavascriptObjectTest.scala]: Typo: "return" => "returns"

**Comment by TheBizzle** [comment]
First, thanks for cleaning up a bunch of little things in the code while you were at this.  It's good to see some of the time-passage residue being swept up.

As for my 95 Theses, unfortunately, on GitHub, it's not easy for me to comment on the exact lines of interest, since the diffs got cut off, thanks to the massive amount of model dump diffs.  Consequently, I'll just comment here, and hopefully it will be easy to find everything.

  * Don't worry about lining up arrows of multiline match clauses (in general, but particularly in `WidgetCompiler`).  I've attempted that one before.  It ultimately doesn't pay.
  * `WidgetCompiler`
    * `compileWidget`: The body should be indented
    * `askWithKind`: `fail` is not a partial function.  It is a `String => Nothing`
  * `JavascriptObject`
    * I managed to figure it out, but a comment on the need for `JavascriptObject` would be good—just something above the class definition that explains the problem this is solving, and why a normal JSON serialization library doesn't cut it.
    * Using `Either` in `JavascriptObject` works for the time being, but only because there are currently only two options.  I would suggest just creating a typeclass for everything that can reside within `JavascriptObject`, and then do away with the `Either` stuff.
      * This should also allow you to do away with the secondary constructors (which I think shouldn't be there, anyway, since, if someone wants a `JavascriptObject` with one function, they should do `new JavascriptObject.addFunction(s, f)`.  Even so, though, you can just make the primary constructor into `class JavascriptObject(elements: JavascriptElement*)` (where `JavascriptElement` is the typeclass I'm proposing) and unify the whole thing.
    * I dislike the names of the methods `asJsonObject` and `asJsonFunctions`, since they (falsely) imply that the operation isn't lossy.  I think that renaming `asJsonFunctions` to `functions` would be sufficient.  As for `asJsonObject`, I could live with a name like `asPrunedJsonObject`, or really anything else that suggests that some of the data passed into the `JavascriptObject` might be lost by calling that method.
    * `asJsonFunctions`: The current definition is `elements.flatMap(_.right.toOption)`.  This would be better phrased as `elements.filter(_.isRight)`.  A similar thing can be said of the definition of `asJsonObject`.
     * If we want to talk programming rhetoric, `flatMap` is the most-general function in existence that holds the powers of the divine.  I recommend first looking for weaker, more-specific combinators before reaching for `flatMap`.
    * `apply`: It would make more sense to me to use `.reverse.find(f)` than `flatMap(f).lastOption`

**Comment by TheBizzle** [comment]
> I was wondering whether it might be better to have the model split into sections, into which each piece could add bits.  [...] If we would like to see a change similar to this one, I'd be happy to open an issue.

Please do.  I agree that the current solution is kind of hairy.

**Comment by mrerrormessage** [comment]
@TheBizzle , this should be ready for review now that I've reworked the javascript aggregation process. The [corresponding branch of Galapagos](https://github.com/NetLogo/Galapagos/pull/244) is also working and is good to review anytime as well.

File: js/src/main/scala/BrowserCompiler.scala [hunk 2]
```diff
@@ -187,62 +161,25 @@ object BrowserCompiler {
...
-  case class SliderCompilation(compiledMin: CompiledStringV,
-    compiledMax: CompiledStringV,
-    compiledStep: CompiledStringV) extends WidgetCompilation
-
   case class ModelCompilation(
     result: String,
     code: String,
     info: String,
-    widgets: Seq[CompiledWidget],
+    widgets: List[JsObject],
```
Comment (TheBizzle) [inline js/src/main/scala/BrowserCompiler.scala]: Is there a reason to want this to be a `List`, rather than the more-general `Seq`?

Comment (SethTisue) [inline js/src/main/scala/BrowserCompiler.scala]: a `Seq` might be mutable (until https://issues.scala-lang.org/browse/SI-6570 gets fixed)

Comment (TheBizzle) [inline js/src/main/scala/BrowserCompiler.scala]: And whose fault is that? :stuck_out_tongue:

**Comment by TheBizzle** [comment]
Please put model dump updates in a separate commit.  They are very noisy, and GitHub hides anything that comes after them, rendering me unable to comment on them directly in the "diff" view.

  * https://github.com/NetLogo/Tortoise/blob/wip-test-monitors-docked/tortoise/src/main/scala/CompiledModel.scala#L24-L27
    * I would recommend just using a destructuring bind.  (e.g. `val Compilation(_, widgets, _, model, procedures, program) = compilation`)
  * https://github.com/NetLogo/Tortoise/blob/wip-test-monitors-docked/tortoise/src/main/scala/CompilerLike.scala#L15-L16
    * The formatting is strange.  Line 16 should be indented more.
  * https://github.com/NetLogo/Tortoise/blob/wip-test-monitors-docked/tortoise/src/main/scala/Compiler.scala#L53-L54
    * Same as above.
  * https://github.com/NetLogo/Tortoise/blob/wip-test-monitors-docked/tortoise/src/main/scala/JavascriptObject.scala#L43
    * This method is never used, nor would its type make sense to me, even if it were.  Avoiding `Either` and just pattern matching on the `Element` at the `get` call site would accomplish the same thing.  Even if you needed something other than `this.type` for some hypothetical `get` method, it could be accomplished by simply making `Element` take a type parameter.

**Comment by TheBizzle** [comment]
Correction to the final point in my preceding comment: I was **wrong** in suggesting that `get` always essentially returns a `this.type`, but it **could** still be accomplished better with a type parameter (in our hypothetical world where `get` is used).

File: js/src/main/scala/BrowserCompiler.scala (hunk 2 above)
Comment (mrerrormessage) [inline js/src/main/scala/BrowserCompiler.scala]: I'll change it to a `Seq`. I think the reason for this was just that `widgets` in [NetLogo Headless `core.model`](https://github.com/NetLogo/NetLogo-Headless/blob/master/parser-core/src/main/core/Model.scala#L5) are contained in a `List`, and I happened to be flipping through that code a lot. I'll change it to a `Seq`. @SethTisue can change `Seq`s to be immutable, everyone wins.

**Comment by mrerrormessage** [comment]
I'll have those changes pushed up sometime tomorrow morning. Thanks for the review.

File: js/src/main/scala/BrowserCompiler.scala (hunk 2 above)
Comment (SethTisue) [inline js/src/main/scala/BrowserCompiler.scala]: I'll put it at the top of my 2.13 wishlist. (Which I only just started, so it's also at the bottom.)

Comment (SethTisue) [inline js/src/main/scala/BrowserCompiler.scala]: fwiw, in my own code my current tendency is to use `Vector` in stuff that isn't API, and grit my teeth a little and use `Seq` in APIs

**Comment by mrerrormessage** [comment]
Comments addressed. The comment on `BrowserCompiler` made me realize that we were returning widget objects from `BrowserCompiler` instead of a string that can be `eval`'ed to give a list of widgets. I have incorporated that into this changeset so that we keep the scala jvm and scala js versions in sync.

**Comment by TheBizzle** [comment]
Once Travis signs off, this will be mergeable.

**Comment by mrerrormessage** [comment]
Travis is failing, but I'm working on a fix right now, just a quick adjustment in `netLogoWeb/test` for the changed format from `BrowserCompiler`.

**Comment by mrerrormessage** [comment]
Travis looks to be passing again. This should be ready for merge. [These were the relevant changes in netlogo-web tests](https://github.com/NetLogo/Tortoise/commit/65c2e3cf29f42d2dd7b0843cdf091aa8d2c71cb8#diff-1c1e5d4c79db9cd2da6fdef6635774dfL27).

**Comment by TheBizzle** [comment]
The refactoring of `RuntimeInit` in particular was much needed.  Thanks!

**Pull request closed without merging**